calliope/preprocess/model_data.py

```diff
@@ -51,325 +55,398 @@ def build_model_data(model_run, debug=False):
-                                        "techs": _tech,
-                                        "carriers": tech_param_info,
-                                    },
-                                ),
-                                1,
-                            )
-                    else:
-                        carrier_tier = tech_param.split(".")[-1].replace("carrier_", "")
-                        if tech_param_info == "resource" or carrier_tier == "carrier":
-                            continue
-                        for _tech in techs:
-                            if not isinstance(tech_param_info, list):
-                                tech_param_info = [tech_param_info]
-                            for _tech_param_info in tech_param_info:
-                                param_dict.set_key(
-                                    set_idx(
-                                        param="carrier",
-                                        keydict={
-                                            "techs": _tech,
-                                            "carriers": _tech_param_info,
-                                            "carrier_tiers": carrier_tier,
-                                        },
-                                    ),
-                                    1,
-                                )
-                else:
-                    param_name = tech_param.split(".")[-1]
-                    set_tech_info(param_dict, techs, param_name, tech_param_info)
-            elif tech_param.startswith("constraints") and "systemwide" in tech_param:
-                param_name = tech_param.split(".")[-1]
-                set_tech_info(param_dict, techs, param_name, tech_param_info)
 
+    def __init__(self, model_run_dict):
+        self.node_dict = model_run_dict.nodes.as_dict_flat()
+        self.tech_dict = model_run_dict.techs.as_dict_flat()
 
-def set_idx(param, keydict):
-    """
-    Set a key:value pair in the re-processed model data dictionary
+        self.model_data = xr.Dataset(
+            coords={"timesteps": model_run_dict.timeseries_data.index}
+        )
+        self.add_attributes(model_run_dict)
+        self.lookup_str = "[\\w\\-]*"  # all alphanumerics + `_` and `-`
+        self.template_config = AttrDict.from_yaml(
+            os.path.join(
+                os.path.dirname(calliope.__file__), "config", "model_data_lookup.yaml"
+            )
+        )
+        self.unwanted_tech_keys = [
+            "allowed_constraints",
+            "required_constraints",
+            "allowed_group_constraints",
+            "allowed_costs",
+            "allowed_switches",
+            "constraints",
+            "essentials.carrier",
+        ]
+
+        self.strip_unwanted_keys()
+        self.add_node_tech_sets()
+
+        if model_run_dict.get_key("model.random_seed", None):
+            np.random.seed(seed=model_run_dict.model.random_seed)
+
+    def empty_or_invalid(self, var):
```

Comment:
I'd mark all methods that are supposed to be private with an underscore: `def _empty_or_invalid(self, var)`.

In fact, `ModelData` has no public methods apart from the default methods of classes in Python.